pkg/controllers/cloudprovideraccount/controller.go

```diff
@@ -183,36 +196,48 @@ func (r *CloudProviderAccountReconciler) getCpaBySecret(s types.NamespacedName)
 			}
 		}
 	}
-	return nil, cpaItems
+	return cpaItems, nil
 }
 
 // watchSecret watch the Secret objects.
 func (r *CloudProviderAccountReconciler) watchSecret() {
+	logWithSecret := r.Log.WithName("Secret")
 	for {
 		event, ok := <-r.watcher.ResultChan()
 		if !ok {
 			r.resetSecretWatcher()
 		} else {
+			secret := event.Object.(*v1.Secret)
+			namespacedName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
 			switch event.Type {
 			case watch.Modified:
 				r.mutex.Lock()
-				secret := event.Object.(*v1.Secret)
-				namespacedName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
-				r.Log.WithName("Secret").Info("Received request", "Secret", namespacedName, "operation", "update")
-				err, cpaItems := r.getCpaBySecret(namespacedName)
+				logWithSecret.Info("Received request", "Secret", namespacedName, "operation", "update")
+				cpaItems, err := r.getCpaBySecret(namespacedName)
 				if err != nil {
 					r.Log.WithName("Secret").Error(err, "error getting CPA by Secret", "account", namespacedName)
 				}
 				for _, cpa := range cpaItems {
 					accountNamespacedName := types.NamespacedName{Namespace: cpa.Namespace, Name: cpa.Name}
-					err := r.processCreateOrUpdate(&accountNamespacedName, &cpa)
-					if err != nil {
-						r.Log.WithName("Secret").Error(err, "error updating account", "account", accountNamespacedName)
+					if err := r.processCreateOrUpdate(&accountNamespacedName, &cpa); err != nil {
+						logWithSecret.Error(err, "error updating account", "account", accountNamespacedName)
 					} else {
-						r.Log.WithName("Secret").Info("Done processing secret", "account", accountNamespacedName)
+						logWithSecret.Info("Done processing update", "account", accountNamespacedName)
 					}
 				}
 				r.mutex.Unlock()
+			case watch.Deleted:
+				r.mutex.Lock()
+				logWithSecret.Info("Received request", "Secret", namespacedName, "operation", "delete")
+				cpaItems, err := r.getCpaBySecret(namespacedName)
+				if err != nil {
+					logWithSecret.Error(err, "error getting CPA by Secret", "account", namespacedName)
+				}
+				for _, cpa := range cpaItems {
+					accountNamespacedName := types.NamespacedName{Namespace: cpa.Namespace, Name: cpa.Name}
+					logWithSecret.Info("Done processing delete", "account", accountNamespacedName)
```

Comment:
info log is not clear.. you have to say, secret object is getting removed, however there is CPA referencing to it. Please delete it.. Also, not sure, if you need locks here.

Reply:
I kept the lock here to make sure when watcher would be doing one operation at a time.  On a second thought, lock is not required, since everything is being processing inline. I will remove the lock.

Reply:
is this better?
```suggestion
										logWithSecret.Info("Done processing Secret delete, update/delete stale account", "account", accountNamespacedName)

```